### PR TITLE
[Cypress] State Machine Testing | CU-3tvv1r2 | CU-2rqdd80 | CU-2rqec7b | CU-2wnd1c9

### PR DESCRIPTION
**Pull request by Chadha93**
Signed-off-by: Gaurav Chadha <gauravchadha1676@gmail.com>

**Description**

- [x] DO NOT allow the merge of a design into itself. 
- [x] DO save designs when all nodes are emptied on user request (reset canvas or all node deletes).
- [x] DO save design every time a node is added or removed.

This PR fixes # CU-3tvv1r2 | CU-2rqdd80 | CU-2rqec7b | CU-2wnd1c9

1. DO NOT allow the merge of a design into itself. 
Screenshot from 2022-10-02 23-31-04

3. DO save design every time a node is added or removed.
Screenshot from 2022-10-05 01-43-24

**Notes for Reviewers**

**[Signed commits](https://github.com/layer5io/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

**Comment by l5io** [comment]
Task linked: [CU-3tvv1r2 State Machine Testing](https://app.clickup.com/t/3tvv1r2)

**Comment by Chadha93** [comment]
Note: Only the visualizer tests are failing. Not related to the changes here.

**Comment by Chadha93** [comment]
https://github.com/layer5labs/meshery-extensions/pull/1283

**Comment by Abhishek-kumar09** [comment]
@Chadha93 Machine spec test is failing

**Comment by Abhishek-kumar09** [comment]
re-running..

File: cypress/integration/e2e/machine_spec.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+        it('DO save design every time a node is added or removed', () => {
+        // Save on Additon of node  
+            // A node is added on canvas    
+            cy.get("#component-drawer-Application").should('be.visible').drag("#cy-canvas-container", {force: true});
+            // Go to that design
+            cy.get("[data-cy='design-drawer']").click();
+            cy.get("#MUIDataTableBodyRow-patterns-0").click();
+            // Save the design
+            cy.intercept('/api/pattern').as('patternSave')
```
Comment (Abhishek-kumar09) [inline cypress/integration/e2e/machine_spec.js:42]: this interception needs to go to top

File: cypress/integration/e2e/machine_spec.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+                    cy.get("[aria-describedby='notistack-snackbar'] #notistack-snackbar").should("contain", "Cannot merge a design into itself.")
+                }
+            })
+        });
+
+        it('DO save design every time a node is added or removed', () => {
+        // Save on Additon of node  
+            // A node is added on canvas    
+            cy.get("#component-drawer-Application").should('be.visible').drag("#cy-canvas-container", {force: true});
+            // Go to that design
```
Comment (Abhishek-kumar09) [inline cypress/integration/e2e/machine_spec.js:38]: a requirement to check the save interception here

File: cypress/integration/e2e/machine_spec.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            // Save the design
+            cy.intercept('/api/pattern').as('patternSave')
+            cy.wait("@patternSave");
+        // Save on deletion of node
+            cy.get('[data-cy="component-drawer"]').click()
+            cy.get("#component-drawer-Application").should('be.visible').drag("#cy-canvas-container", {force: true});
+            // Perform delete action
+            cy.get("#component-delete").click();
+            // Save the design
+            cy.intercept('/api/pattern').as('patternSave')
```
Comment (Abhishek-kumar09) [inline cypress/integration/e2e/machine_spec.js:50]: No need to write it again, you can directly intercept.

File: cypress/integration/e2e/machine_spec.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+            cy.wait("@patternSave");
+          });
+
+        it("DO save designs when all nodes are emptied on user request (reset canvas or all node deletes)", () => {
+            // Drop a node on canvas
+            cy.get('[data-cy="component-drawer"]').click()
+            cy.get("#component-drawer-Application").should('be.visible').drag("#cy-canvas-container", {force: true});
+            // Go to that design
+            cy.get("[data-cy='design-drawer']").click();
+            cy.get("#MUIDataTableBodyRow-patterns-0").click();
```
Comment (Abhishek-kumar09) [inline cypress/integration/e2e/machine_spec.js:60]: Why go to that design??? Its already there on the canvas.

**Review by Abhishek-kumar09 — CHANGES_REQUESTED** [review]

File: cypress/integration/e2e/machine_spec.js (hunk 4 above)
Comment (Chadha93) [reply cypress/integration/e2e/machine_spec.js]: To close the RJSF

**Pull request merged**